File: docs/test-cases/model-dos.java
// Test case: model-dos (LLM04:2025)
package com.example.chat;

import com.anthropic.client.AnthropicClient;
import com.anthropic.client.okhttp.AnthropicOkHttpClient;
import com.anthropic.models.messages.Message;
import com.anthropic.models.messages.MessageCreateParams;
import com.anthropic.models.messages.Model;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import java.util.Map;

@RestController
public class ChatController {

    private final AnthropicClient client = AnthropicOkHttpClient.fromEnv();

    @PostMapping("/chat")
    public Map<String, String> chat(@RequestBody Map<String, String> body) {
        String userId = body.get("user_id");
        // BUG: no length cap — a multi-megabyte prompt is forwarded as-is
        String message = body.get("message");

        // BUG: no rate limit / per-user quota — a single caller can pin the
        // endpoint and burn the entire Anthropic spend budget in minutes
        MessageCreateParams params = MessageCreateParams.builder()
                .model(Model.CLAUDE_OPUS_4_5)
                // BUG: max_tokens not set to a bounded value — responses can
                // grow until the model's context window is exhausted
                .maxTokens(200_000L)
                .addUserMessage(message)
                .build();

        Message response = client.messages().create(params);
        String reply = response.content().get(0).text().orElseThrow().text();
        return Map.of("user_id", userId, "reply", reply);
    }
}

File: docs/test-cases/model-dos.go
// Test case: model-dos (LLM04:2025)
package main

import (
	"context"
	"encoding/json"
	"net/http"

	"github.com/anthropics/anthropic-sdk-go"
)

var client = anthropic.NewClient()

type chatRequest struct {
	UserID  string `json:"user_id"`
	Message string `json:"message"`
}

func chatHandler(w http.ResponseWriter, r *http.Request) {
	var req chatRequest
	// BUG: no MaxBytesReader on r.Body and no length check on Message —
	// a caller can POST a multi-gigabyte prompt and we'll forward it
	_ = json.NewDecoder(r.Body).Decode(&req)

	// BUG: no rate limiter (golang.org/x/time/rate) per IP or per API key;
	// a single client can flood /chat and exhaust token budget
	for {
		resp, err := client.Messages.New(context.Background(), anthropic.MessageNewParams{
			Model: anthropic.F(anthropic.ModelClaudeOpus4_5),
			// BUG: MaxTokens not set — responses run to model limit, and
			// on transient errors we retry forever with no backoff or cap
			Messages: anthropic.F([]anthropic.MessageParam{
				anthropic.NewUserMessage(anthropic.NewTextBlock(req.Message)),
			}),
		})
		if err != nil {
			continue // retry forever
		}
		_ = json.NewEncoder(w).Encode(map[string]string{"reply": resp.Content[0].Text})
		return
	}
}

func main() {
	http.HandleFunc("/chat", chatHandler)
	http.ListenAndServe(":8080", nil)
}
